### PR TITLE
[Example] Add VAE example

### PR DESCRIPTION
**Pull request by DrRyanHuang**
### PR types

New features

### PR changes

Others

### Describe

目前约完成 1/2 还在精度对齐ing

- 参考案例:  https://aistudio.baidu.com/projectdetail/5541961
- https://github.com/PaddlePaddle/PaddleScience/issues/440

### CODE REVIEW

**Comment by HydrogenSulfate** [comment]
感谢大佬提PR，可以合并一下最新develop分支的代码，然后案例可以用hydra改造一下，参考这个案例：https://github.com/PaddlePaddle/PaddleScience/tree/develop/examples/bracket

**Comment by DrRyanHuang** [comment]
@HydrogenSulfate 不好意思，打扰了，AISTUDIO原文用到了 `Julia` 环境，我这边写test的时候也配置配置一个 `Julia` 环境，还是怎么操作一下嘞? 😆

原文地址 ：https://aistudio.baidu.com/projectdetail/5541961

```python
# julia 依赖
os.environ['JULIA_DEPOT_PATH'] = '/home/aistudio/opt/julia_package'
# pip 依赖
sys.path.append('/home/aistudio/opt/external-libraries')

# julieries
from julia.api import Julia
jl = Julia(compiled_modules=False,runtime="/home/aistudio/opt/julia-1.8.5/bin/julia")
# import julia
from julia import Main
```

**Comment by HydrogenSulfate** [comment]
> @HydrogenSulfate 不好意思，打扰了，AISTUDIO原文用到了 `Julia` 环境，我这边写test的时候也配置配置一个 `Julia` 环境，还是怎么操作一下嘞? 😆
> 
> 原文地址 ：https://aistudio.baidu.com/projectdetail/5541961
> 
> ```python
> # julia 依赖
> os.environ['JULIA_DEPOT_PATH'] = '/home/aistudio/opt/julia_package'
> # pip 依赖
> sys.path.append('/home/aistudio/opt/external-libraries')
> 
> # julieries
> from julia.api import Julia
> jl = Julia(compiled_modules=False,runtime="/home/aistudio/opt/julia-1.8.5/bin/julia")
> # import julia
> from julia import Main
> ```



> @HydrogenSulfate 不好意思，打扰了，AISTUDIO原文用到了 `Julia` 环境，我这边写test的时候也配置配置一个 `Julia` 环境，还是怎么操作一下嘞? 😆
> 
> 原文地址 ：https://aistudio.baidu.com/projectdetail/5541961
> 
> ```python
> # julia 依赖
> os.environ['JULIA_DEPOT_PATH'] = '/home/aistudio/opt/julia_package'
> # pip 依赖
> sys.path.append('/home/aistudio/opt/external-libraries')
> 
> # julieries
> from julia.api import Julia
> jl = Julia(compiled_modules=False,runtime="/home/aistudio/opt/julia-1.8.5/bin/julia")
> # import julia
> from julia import Main
> ```

测试文件test.ipynb只是一个测试功能是否正常的文件，本地测试使用，可以不用提交到PR里。

**Comment by HydrogenSulfate** [comment]
> @HydrogenSulfate 不好意思，打扰了，AISTUDIO原文用到了 `Julia` 环境，我这边写test的时候也配置配置一个 `Julia` 环境，还是怎么操作一下嘞? 😆
> 
> 原文地址 ：https://aistudio.baidu.com/projectdetail/5541961
> 
> ```python
> # julia 依赖
> os.environ['JULIA_DEPOT_PATH'] = '/home/aistudio/opt/julia_package'
> # pip 依赖
> sys.path.append('/home/aistudio/opt/external-libraries')
> 
> # julieries
> from julia.api import Julia
> jl = Julia(compiled_modules=False,runtime="/home/aistudio/opt/julia-1.8.5/bin/julia")
> # import julia
> from julia import Main
> ```

可以将上述路径配置写到hydra yaml文件里，然后通过cfg.JULIA.xxx访问即可。

File: examples/RegAE/train_new.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+from typing import Dict
+from typing import List
+
+from paddle import nn
+from paddle.nn import functional as F
+import sys; sys.path.append(r"C:\Users\zihao\Desktop\lbwnb\PaddleScience")
```
Comment (HydrogenSulfate) [inline examples/RegAE/train_new.py:23]: 这行可以删除，因为通过 `pip install -e .` 或 `export PYTHONPATH=\`pwd\`` 的方式安装，应该都不需要这行路径设置代码。

File: examples/RegAE/train_new.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    import pgl
+
+
+criterion = nn.MSELoss()
+kl_loss = KLLoss01()
+    
+# def train_mse_func(
+#     output_dict: Dict[str, "paddle.Tensor"], label_dict: Dict[str, "pgl.Graph"], *args
+# ) -> paddle.Tensor:
+#     return F.mse_loss(output_dict["pred"], label_dict["label"].y)
```
Comment (HydrogenSulfate) [inline examples/RegAE/train_new.py:40]: 一些无用的注释可以删除

File: examples/RegAE/train_new.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    # initialize logger
+    logger.init_logger("ppsci", f"{OUTPUT_DIR}/train.log", "info")
+
+    latent_dim, hidden_dim = 100, 100
+    # set model
+    model = ppsci.arch.AutoEncoder(
+        input_dim=10000, 
+        latent_dim=latent_dim, 
+        hidden_dim=hidden_dim,
+    )
```
Comment (HydrogenSulfate) [inline examples/RegAE/train_new.py:77]: AutoEncoder的初始化超参数移动至RegAE.yaml配置文件里

File: examples/RegAE/train_new.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        ITERS_PER_EPOCH,
+        save_freq=50,
+        eval_during_train=True,
+        eval_freq=50,
+        # validator=validator,
+        eval_with_no_grad=True,
+        # pretrained_model_path="./output_AMGNet/checkpoints/latest"
+    )
+    # train model
+    solver.train()
```
Comment (HydrogenSulfate) [inline examples/RegAE/train_new.py:152]: 文件末尾换行

File: ppsci/arch/vae.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            nn.Tanh(),
+        )
+        self._encoder_mu = nn.Linear(hidden_dim, latent_dim)
+        self._encoder_log_sigma = nn.Linear(hidden_dim, latent_dim)
+
+        self._decoder = nn.Sequential(
+            nn.Linear(latent_dim, hidden_dim),
+            nn.Tanh(),
+            nn.Linear(hidden_dim, input_dim),
+        )
```
Comment (HydrogenSulfate) [inline ppsci/arch/vae.py:46]: base.Arch的子类需要做成符号化的形式，即需要具有 input_keys、output_keys两个属性

File: ppsci/arch/vae.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        return self._decoder(x)
+
+    # @staticmethod
+    # def kl_loss(mu, log_sigma):
+    #     # 计算mu，log_sigma与 N(0,1)分布的差距
+    #     base = paddle.exp(2. * log_sigma) + paddle.pow(mu, 2) - 1. - 2. * log_sigma
+    #     loss = 0.5 * paddle.sum(base) / mu.shape[0]
+    #     return loss
+
+    def forward(self, x, noise):
```
Comment (HydrogenSulfate) [inline ppsci/arch/vae.py:64]: forward中的张量计算放到forward_tensor中，然后forward内主要负责从输入的字典中找到并组装出真正的输入张量

File: ppsci/data/dataset/npz_dataset.py [hunk 7]
```diff
@@ -265,3 +265,80 @@ def __iter__(self):
...
+class ScalerStd(object):
+    """
+    Desc: Normalization utilities with std mean
+    """
+
+    def __init__(self):
+        self.mean = 0.
+        self.std = 1.
+
+    def fit(self, data):
```
Comment (HydrogenSulfate) [inline ppsci/data/dataset/npz_dataset.py:282]: 内部使用的函数前面加一个下划线：`_fit`

File: ppsci/data/dataset/npz_dataset.py [hunk 8]
```diff
@@ -265,3 +265,80 @@ def __iter__(self):
...
+        self.mean = np.mean(data)
+        self.std = np.std(data)
+
+    def transform(self, data):
+        mean = paddle.to_tensor(self.mean).type_as(data).to(data.device) if paddle.is_tensor(data) else self.mean
+        std = paddle.to_tensor(self.std).type_as(data).to(data.device) if paddle.is_tensor(data) else self.std
+        return (data - mean) / std
+
+    def inverse_transform(self, data):
```
Comment (HydrogenSulfate) [inline ppsci/data/dataset/npz_dataset.py:291]: 同上

File: ppsci/data/dataset/npz_dataset.py [hunk 9]
```diff
@@ -265,3 +265,80 @@ def __iter__(self):
 
     def __len__(self):
         return 1
+
+
+
+
+
+class ScalerStd(object):
```
Comment (HydrogenSulfate) [inline ppsci/data/dataset/npz_dataset.py:273]: 删除(object)，python3已经不需要从object继承

File: ppsci/data/dataset/npz_dataset.py [hunk 10]
```diff
@@ -265,3 +265,80 @@ def __iter__(self):
...
+        self.test_data = self.scaler.transform(self.test_data)
+        
+        self.input_keys = ""
+        self.label_keys = ""
+
+    def __getitem__(self, idx):
+        if self.data_type == "train":
+            return self.train_data[idx]
+        else:
+            return self.test_data[idx]
```
Comment (HydrogenSulfate) [inline ppsci/data/dataset/npz_dataset.py:338]: dataset需要返回input、label、weight三个数据字典

File: ppsci/loss/kl.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    #     super().__init__(reduction, weight)
+
+    def __init__(self):
+        super().__init__(None, None)
+
+    def forward(self, mu, log_sigma):
+        # 计算mu，log_sigma与 N(0,1)分布的差距
+        base = paddle.exp(2. * log_sigma) + paddle.pow(mu, 2) - 1. - 2. * log_sigma
+        loss = 0.5 * paddle.sum(base) / mu.shape[0]
+        return loss
```
Comment (HydrogenSulfate) [inline ppsci/loss/kl.py:50]: 文件末尾换行

File: ppsci/loss/kl.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    #     if reduction not in ["mean", "sum"]:
+    #         raise ValueError(
+    #             f"reduction should be 'mean' or 'sum', but got {reduction}"
+    #         )
+    #     super().__init__(reduction, weight)
+
+    def __init__(self):
+        super().__init__(None, None)
+
+    def forward(self, mu, log_sigma):
```
Comment (HydrogenSulfate) [inline ppsci/loss/kl.py:46]: 所有loss子类的forward写法统一，接受output_dict, label_dict, weight_dict三个变量，然后计算loss并以一个标量Tensor的形式返回。可以将N(0,1)的数据作为标签数据，从label_dict传入

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

**Comment by HydrogenSulfate** [comment]
closed for #660
